### PR TITLE
Fix MakeStateful transformation: use tensor names instead of friendly names

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 - *MakeStateful transformation: Use tensor names instead of friendly names, handle one output tensor to several Result ops case*

### Tickets:
 - *68304*

### CODE REVIEW

File: src/core/src/pass/make_stateful.cpp [hunk 1]
```diff
@@ -26,22 +26,45 @@ ov::pass::MakeStateful::ParamResPairs find_param_results_by_names(
...
     for (const auto& param_res : param_res_names) {
         const auto& param_name = param_res.first;
         const auto& res_name = param_res.second;
-
         auto param = std::find_if(params.begin(), params.end(), [&](const std::shared_ptr<ngraph::Node>& node) {
-            return node->get_friendly_name() == param_name;
+            const auto& possible_names = node->output(0).get_names();
+            return possible_names.find(param_name) != possible_names.end();
         });
         NGRAPH_CHECK(param != params.end(), "Parameter node with name = ", param_name, "doesn't exist in the function");
```
Comment (GlebKazantaev) [inline src/core/src/pass/make_stateful.cpp:40]: Please update exception message.

File: src/core/src/pass/make_stateful.cpp [hunk 2]
```diff
@@ -26,22 +26,45 @@ ov::pass::MakeStateful::ParamResPairs find_param_results_by_names(
...
+
         NGRAPH_CHECK(res != results.end(), "Result node with name = ", res_name, " doesn't exist in the function");
 
-        pairs_to_replace.emplace_back(*param, *res);
+        // In case of several Results connected to one output tensor,
+        // We can't determine what result we need to take exactly.
+        // But we can take first unused, the order is not important, data is the same.
+        std::shared_ptr<opset8::Result> unused_res;
+        for (const auto& target_in : (*res)->input_value(0).get_target_inputs()) {
+            auto is_target_res = std::dynamic_pointer_cast<opset8::Result>(target_in.get_node()->shared_from_this());
```
Comment (GlebKazantaev) [inline src/core/src/pass/make_stateful.cpp:55]:
```suggestion
            auto is_target_res = is_type<opset8::Result>(target_in.get_node());
```

File: src/core/src/pass/make_stateful.cpp [hunk 3]
```diff
@@ -26,22 +26,45 @@ ov::pass::MakeStateful::ParamResPairs find_param_results_by_names(
...
+            if (!is_target_res) {
+                continue;
+            }
+            if (uniq_res.find(is_target_res.get()) == uniq_res.end()) {
```
Comment (GlebKazantaev) [inline src/core/src/pass/make_stateful.cpp:59]:
```suggestion
            if (uniq_res.find(is_target_res) == uniq_res.end()) {
```

File: src/core/src/pass/make_stateful.cpp [hunk 4]
```diff
@@ -26,22 +26,45 @@ ov::pass::MakeStateful::ParamResPairs find_param_results_by_names(
...
+                unused_res = is_target_res;
+                break;
+            }
+        }
+        NGRAPH_CHECK(unused_res != nullptr, "");
```
Comment (GlebKazantaev) [inline src/core/src/pass/make_stateful.cpp:64]: exception message?

File: src/core/src/pass/make_stateful.cpp (hunk 1 above)
Comment (itikhono) [reply src/core/src/pass/make_stateful.cpp]: done

File: src/core/src/pass/make_stateful.cpp (hunk 3 above)
Comment (itikhono) [reply src/core/src/pass/make_stateful.cpp]: fixed

File: src/core/src/pass/make_stateful.cpp (hunk 4 above)
Comment (itikhono) [reply src/core/src/pass/make_stateful.cpp]: added

File: src/core/src/pass/make_stateful.cpp (hunk 2 above)
Comment (itikhono) [reply src/core/src/pass/make_stateful.cpp]: done

**Review by GlebKazantaev — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**